pkgs/tools/text/pyp/default.nix

```diff
@@ -0,0 +1,47 @@
+{ lib
+, fetchFromGitHub
+, python3
+, bc
+, jq
+}:
+
+let
+  version = "1.1.0";
+in python3.pkgs.buildPythonApplication {
+  pname = "pyp";
+  inherit version;
+  format = "pyproject";
+
+  src = fetchFromGitHub {
+    owner = "hauntsaninja";
+    repo = "pyp";
+    rev = "v${version}";
+    hash = "sha256-A1Ip41kxH17BakHEWEuymfa24eBEl5FIHAWL+iZFM4I=";
+  };
+
+  nativeBuildInputs = [
+    python3.pkgs.flit-core
+  ];
+
+  nativeCheckInputs = [
+    python3.pkgs.pytest
+    bc
+    jq
+  ];
+
+  checkPhase = ''
+    runHook preCheck
+
+    PATH=$out/bin:$PATH pytest
+
+    runHook postCheck
+  '';
```

Comment:
Strangely, no: it does not find pyp executable.

> pyp> Sourcing python-remove-tests-dir-hook
> pyp> Sourcing python-catch-conflicts-hook.sh
> pyp> Sourcing python-remove-bin-bytecode-hook.sh
> pyp> Sourcing pip-build-hook
> pyp> Using pipBuildPhase
> pyp> Using pipShellHook
> pyp> Sourcing pip-install-hook
> pyp> Using pipInstallPhase
> pyp> Sourcing python-imports-check-hook.sh
> pyp> Using pythonImportsCheckPhase
> pyp> Sourcing python-namespaces-hook
> pyp> Sourcing python-catch-conflicts-hook.sh
> pyp> Sourcing pytest-check-hook
> pyp> Using pytestCheckPhase
> pyp> unpacking sources
> pyp> unpacking source archive /nix/store/9ydh1p0pa5284jrmkf2km0f1wi83xngz-source
> pyp> source root is source
> pyp> setting SOURCE_DATE_EPOCH to timestamp 315619200 of file source/tox.ini
> pyp> patching sources
> pyp> updateAutotoolsGnuConfigScriptsPhase
> pyp> configuring
> pyp> no configure script, doing nothing
> pyp> building
> pyp> Executing pipBuildPhase
> pyp> Creating a wheel...
> pyp> WARNING: The directory '/homeless-shelter/.cache/pip' or its parent directory is not owned or is not writable by the current user. The cache has been disabled. Check the permissions and owner of that directory. If executing pip with sudo, you should use sudo's -H flag.
> pyp> Processing /build/source
> pyp>   Running command Preparing metadata (pyproject.toml)
> pyp>   Preparing metadata (pyproject.toml) ... done
> pyp> Building wheels for collected packages: pypyp
> pyp>   Running command Building wheel for pypyp (pyproject.toml)
> pyp>   Building wheel for pypyp (pyproject.toml) ... done
> pyp>   Created wheel for pypyp: filename=pypyp-1.1.0-py3-none-any.whl size=15399 sha256=2cc4e527e98c4345213572b6128cb4283c1e22e65313c3fff785e65c66516782
> pyp>   Stored in directory: /build/pip-ephem-wheel-cache-of7b683u/wheels/79/4b/9b/fc24941d75fdb959a71ee80195b2c89d6ae9389f9f0f80fb22
> pyp> Successfully built pypyp
> pyp> Finished creating a wheel...
> pyp> Finished executing pipBuildPhase
> pyp> installing
> pyp> Executing pipInstallPhase
> pyp> /build/source/dist /build/source
> pyp> Processing ./pypyp-1.1.0-py3-none-any.whl
> pyp> Installing collected packages: pypyp
> pyp> Successfully installed pypyp-1.1.0
> pyp> /build/source
> pyp> Finished executing pipInstallPhase
> pyp> pythonOutputDistPhase
> pyp> Executing pythonOutputDistPhase
> pyp> Finished executing pythonOutputDistPhase
> pyp> post-installation fixup
> pyp> shrinking RPATHs of ELF executables and libraries in /nix/store/v29ag8dwqm8lfqygfi0ffpgx2l4g7cvf-pyp-1.1.0
> pyp> checking for references to /build/ in /nix/store/v29ag8dwqm8lfqygfi0ffpgx2l4g7cvf-pyp-1.1.0...
> pyp> patching script interpreter paths in /nix/store/v29ag8dwqm8lfqygfi0ffpgx2l4g7cvf-pyp-1.1.0
> pyp> stripping (with command strip and flags -S -p) in  /nix/store/v29ag8dwqm8lfqygfi0ffpgx2l4g7cvf-pyp-1.1.0/lib /nix/store/v29ag8dwqm8lfqygfi0ffpgx2l4g7cvf-pyp-1.1.0/bin
> pyp> shrinking RPATHs of ELF executables and libraries in /nix/store/nwyccxf1hjasavcy5zv0wbr92c63hawq-pyp-1.1.0-dist
> pyp> checking for references to /build/ in /nix/store/nwyccxf1hjasavcy5zv0wbr92c63hawq-pyp-1.1.0-dist...
> pyp> patching script interpreter paths in /nix/store/nwyccxf1hjasavcy5zv0wbr92c63hawq-pyp-1.1.0-dist
> pyp> Rewriting #!/nix/store/9c03r86hcdn43dm3hsgjirifvyzfkhwh-python3-3.10.12/bin/python3.10 to #!/nix/store/9c03r86hcdn43dm3hsgjirifvyzfkhwh-python3-3.10.12
> pyp> wrapping `/nix/store/v29ag8dwqm8lfqygfi0ffpgx2l4g7cvf-pyp-1.1.0/bin/pyp'...
> pyp> Executing pythonRemoveTestsDir
> pyp> Finished executing pythonRemoveTestsDir
> pyp> running install tests
> pyp> no Makefile or custom installCheckPhase, doing nothing
> pyp> pythonCatchConflictsPhase
> pyp> pythonRemoveBinBytecodePhase
> pyp> pythonImportsCheckPhase
> pyp> Executing pythonImportsCheckPhase
> pyp> pytestCheckPhase
> pyp> Executing pytestCheckPhase
> pyp> ============================= test session starts ==============================
> pyp> platform linux -- Python 3.10.12, pytest-7.2.1, pluggy-1.0.0
> pyp> rootdir: /build/source
> pyp> collected 38 items                                                             
> pyp> tests/test_find_names.py ...........xx                                   [ 34%]
> pyp> tests/test_pyp.py F..F...................x.                              [100%]
> pyp> =================================== FAILURES ===================================
> pyp> ________________________________ test_examples _________________________________
> pyp>     def test_examples():
> pyp>         """Test approximately the examples in the README."""
> pyp>         example = "python\nsnake\nss\nmagpy\npiethon\nreadme.md\n"
> pyp>         compare_command(example_cmd="cut -c1-4", pyp_cmd="pyp 'x[:4]'", input=example)
> pyp>         compare_command(
> pyp>             example_cmd="wc -c | tr -d ' '", pyp_cmd="pyp 'len(stdin.read())'", input=example
> pyp>         )
> pyp>         compare_command(
> pyp>             example_cmd="awk '{s+=$1} END {print s}' ",
> pyp>             pyp_cmd="pyp 'sum(map(int, lines))'",
> pyp>             input="1\n2\n3\n4\n5\n",
> pyp>         )
> pyp> >       compare_command(
> pyp>             example_cmd="sh",
> pyp>             pyp_cmd="pyp 'subprocess.run(lines[0], shell=True); pass'",
> pyp>             input="echo echo",
> pyp>             use_subprocess=True,
> pyp>         )
> pyp> tests/test_pyp.py:97:
> pyp> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> pyp> tests/test_pyp.py:62: in compare_command
> pyp>     pyp_output = run_cmd(pyp_cmd, input) if use_subprocess else run_pyp(pyp_cmd, input)
> pyp> tests/test_pyp.py:30: in run_cmd
> pyp>     proc = subprocess.run(
> pyp> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
> pyp> input = b'echo echo', capture_output = False, timeout = None, check = True
> pyp> popenargs = ("pyp 'subprocess.run(lines[0], shell=True); pass'",)
> pyp> kwargs = {'shell': True, 'stderr': -2, 'stdin': -1, 'stdout': -1}
> pyp> process = <Popen: returncode: 127 args: "pyp 'subprocess.run(lines[0], shell=True); pa...>
> pyp> stdout = b'/nix/store/a7f7xfp9wyghf44yv6l6fv9dfw492hd3-bash-5.2-p15/bin/sh: line 1: pyp: command not found\n'
> pyp> stderr = None, retcode = 127
> pyp>     def run(*popenargs,
> pyp>             input=None, capture_output=False, timeout=None, check=False, **kwargs):
> pyp>         """Run command with arguments and return a CompletedProcess instance.
> pyp>         The returned instance will have attributes args, returncode, stdout and
> pyp>         stderr. By default, stdout and stderr are not captured, and those attributes
> pyp>         will be None. Pass stdout=PIPE and/or stderr=PIPE in order to capture them,
> pyp>         or pass capture_output=True to capture both.
> pyp>         If check is True and the exit code was non-zero, it raises a
> pyp>         CalledProcessError. The CalledProcessError object will have the return code
> pyp>         in the returncode attribute, and output & stderr attributes if those streams
> pyp>         were captured.
> pyp>         If timeout is given, and the process takes too long, a TimeoutExpired
> pyp>         exception will be raised.
> pyp>         There is an optional argument "input", allowing you to
> pyp>         pass bytes or a string to the subprocess's stdin.  If you use this argument
> pyp>         you may not also use the Popen constructor's "stdin" argument, as
> pyp>         it will be used internally.
> pyp>         By default, all communication is in bytes, and therefore any "input" should
> pyp>         be bytes, and the stdout and stderr will be bytes. If in text mode, any
> pyp>         "input" should be a string, and stdout and stderr will be strings decoded
> pyp>         according to locale encoding, or by "encoding" if set. Text mode is
> pyp>         triggered by setting any of text, encoding, errors or universal_newlines.
> pyp>         The other arguments are the same as for the Popen constructor.
> pyp>         """
> pyp>         if input is not None:
> pyp>             if kwargs.get('stdin') is not None:
> pyp>                 raise ValueError('stdin and input arguments may not both be used.')
> pyp>             kwargs['stdin'] = PIPE
> pyp>         if capture_output:
> pyp>             if kwargs.get('stdout') is not None or kwargs.get('stderr') is not None:
> pyp>                 raise ValueError('stdout and stderr arguments may not be used '
> pyp>                                  'with capture_output.')
> pyp>             kwargs['stdout'] = PIPE
> pyp>             kwargs['stderr'] = PIPE
> pyp>         with Popen(*popenargs, **kwargs) as process:
> pyp>             try:
> pyp>                 stdout, stderr = process.communicate(input, timeout=timeout)
> pyp>             except TimeoutExpired as exc:
> pyp>                 process.kill()
> pyp>                 if _mswindows:
> pyp>                     # Windows accumulates the output in a single blocking
> pyp>                     # read() call run on child threads, with the timeout
> pyp>                     # being done in a join() on those threads.  communicate()
> pyp>                     # _after_ kill() is required to collect that and add it
> pyp>                     # to the exception.
> pyp>                     exc.stdout, exc.stderr = process.communicate()
> pyp>                 else:
> pyp>                     # POSIX _communicate already populated the output so
> pyp>                     # far into the TimeoutExpired exception.
> pyp>                     process.wait()
> pyp>                 raise
> pyp>             except:  # Including KeyboardInterrupt, communicate handled that.
> pyp>                 process.kill()
> pyp>                 # We don't call process.wait() as .__exit__ does that for us.
> pyp>                 raise
> pyp>             retcode = process.poll()
> pyp>             if check and retcode:
> pyp> >               raise CalledProcessError(retcode, process.args,
> pyp>                                          output=stdout, stderr=stderr)
> pyp> E               subprocess.CalledProcessError: Command 'pyp 'subprocess.run(lines[0], shell=True); pass'' returned non-zero exit status 127.
> pyp> /nix/store/9c03r86hcdn43dm3hsgjirifvyzfkhwh-python3-3.10.12/lib/python3.10/subprocess.py:526: CalledProcessError
> pyp> _______________________________ test_tracebacks ________________________________
> pyp>     def test_tracebacks():
> pyp>         # If our sins against traceback implementation details come back to haunt us, and we can't
> pyp>         # reconstruct a traceback, check that we still output something reasonable
> pyp>         TBE = traceback.TracebackException
> pyp>         with patch("traceback.TracebackException") as mock_tb:
> pyp>             count = 0
> pyp>             def effect(*args, **kwargs):
> pyp>                 nonlocal count
> pyp>                 if count == 0:
> pyp>                     assert args[0] == ZeroDivisionError
> pyp>                     count += 1
> pyp>                     raise Exception
> pyp>                 return TBE(*args, **kwargs)
> pyp>             mock_tb.side_effect = effect
> pyp>             pattern = re.compile("Code raised.*ZeroDivisionError", re.DOTALL)
> pyp>             with pytest.raises(pyp.PypError, match=pattern) as e:
> pyp>                 run_pyp("pyp '1 / 0'")
> pyp>             # Make sure that the test works and we couldn't actually reconstruct a traceback
> pyp>             assert "Possible" not in e.value.args[0]
> pyp>         # Check the entire output, end to end
> pyp>         pyp_error = run_cmd("pyp 'def f(): 1/0' 'f()'", check=False)
> pyp>         message = lambda po, pc: (  # noqa
> pyp>             (
> pyp>                 "error: Code raised the following exception, "
> pyp>                 "consider using --explain to investigate:\n\n"
> pyp>                 "Possible reconstructed traceback (most recent call last):\n"
> pyp>                 '  File "<pyp>", in <module>\n'
> pyp>                 "    output = f()\n"
> pyp>             )
> pyp>             + ("     ^^^^^^^^^^^\n" if sys.version_info >= (3, 11) else "")
> pyp>             + ('  File "<pyp>", in f\n' f"    {po}1 / 0{pc}\n")
> pyp>             + ("          \n" if sys.version_info >= (3, 11) else "")
> pyp>             + ("ZeroDivisionError: division by zero\n")
> pyp>         )
> pyp>         print(repr(pyp_error))
> pyp>         print(repr(message("", "")))
> pyp> >       assert pyp_error == message("(", ")") or pyp_error == message("", "")
> pyp> E       assert ('/nix/store/a...d not found\n' == 'error: Code ...ion by zero\n'
> pyp> E         + /nix/store/a7f7xfp9wyghf44yv6l6fv9dfw492hd3-bash-5.2-p15/bin/sh: line 1: pyp: command not found
> pyp> E         - error: Code raised the following exception, consider using --explain to investigate:
> pyp> E         - 
> pyp> E         - Possible reconstructed traceback (most recent call last):
> pyp> E         -   File "<pyp>", in <module>
> pyp> E         -     output = f()
> pyp> E         -   File "<pyp>", in f...
> pyp> E         
> pyp> E         ...Full output truncated (3 lines hidden), use '-vv' to show or '/nix/store/a...d not found\n' == 'error: Code ...ion by zero\n'
> pyp> E         + /nix/store/a7f7xfp9wyghf44yv6l6fv9dfw492hd3-bash-5.2-p15/bin/sh: line 1: pyp: command not found
> pyp> E         - error: Code raised the following exception, consider using --explain to investigate:
> pyp> E         - 
> pyp> E         - Possible reconstructed traceback (most recent call last):
> pyp> E         -   File "<pyp>", in <module>
> pyp> E         -     output = f()
> pyp> E         -   File "<pyp>", in f...
> pyp> E         
> pyp> E         ...Full output truncated (3 lines hidden), use '-vv' to show)
> pyp> tests/test_pyp.py:216: AssertionError
> pyp> ----------------------------- Captured stdout call -----------------------------
> pyp> '/nix/store/a7f7xfp9wyghf44yv6l6fv9dfw492hd3-bash-5.2-p15/bin/sh: line 1: pyp: command not found\n'
> pyp> 'error: Code raised the following exception, consider using --explain to investigate:\n\nPossible reconstructed traceback (most recent call last):\n  File "<pyp>", in <module>\n    output = f()\n  File "<pyp>", in f\n    1 / 0\nZeroDivisionError: division by zero\n'
> pyp> =========================== short test summary info ============================
> pyp> FAILED tests/test_pyp.py::test_examples - subprocess.CalledProcessError: Command 'pyp 'subprocess.run(lines[0], shell...
> pyp> FAILED tests/test_pyp.py::test_tracebacks - assert ('/nix/store/a...d not found\n' == 'error: Code ...ion by zero\n'
> pyp> =================== 2 failed, 33 passed, 3 xfailed in 0.45s ====================
> pyp> /nix/store/fzb9wy1yz0hn69vxw12954szvrjnjjgk-stdenv-linux/setup: line 1596: pop_var_context: head of shell_variables not a function context
> error: builder for '/nix/store/akyakxpjvr57ifpyrxcdshs60ds6caxd-pyp-1.1.0.drv' failed with exit code 1;
>        last 10 log lines:
>        >
>        > tests/test_pyp.py:216: AssertionError
>        > ----------------------------- Captured stdout call -----------------------------
>        > '/nix/store/a7f7xfp9wyghf44yv6l6fv9dfw492hd3-bash-5.2-p15/bin/sh: line 1: pyp: command not found\n'
>        > 'error: Code raised the following exception, consider using --explain to investigate:\n\nPossible reconstructed traceback (most recent call last):\n  File "<pyp>", in <module>\n    output = f()\n  File "<pyp>", in f\n    1 / 0\nZeroDivisionError: division by zero\n'
>        > =========================== short test summary info ============================
>        > FAILED tests/test_pyp.py::test_examples - subprocess.CalledProcessError: Command 'pyp 'subprocess.run(lines[0], shell...
>        > FAILED tests/test_pyp.py::test_tracebacks - assert ('/nix/store/a...d not found\n' == 'error: Code ...ion by zero\n'
>        > =================== 2 failed, 33 passed, 3 xfailed in 0.45s ====================
>        > /nix/store/fzb9wy1yz0hn69vxw12954szvrjnjjgk-stdenv-linux/setup: line 1596: pop_var_context: head of shell_variables not a function context
>        For full logs, run 'nix log /nix/store/akyakxpjvr57ifpyrxcdshs60ds6caxd-pyp-1.1.0.drv'.
>